File: docs/perry-main-src/perry-main/benchmarks/json_polyglot/bench.swift
// JSON parse + stringify polyglot benchmark — Swift (Foundation JSONEncoder/JSONDecoder).
// 10k records, ~1 MB blob, 50 iterations.
// IDENTICAL workload to bench.ts / bench.go / bench.rs / bench.cpp / bench.js.

import Foundation

struct Nested: Codable {
    let x: Int
    let y: Int
}

struct Item: Codable {
    let id: Int
    let name: String
    let value: Double
    let tags: [String]
    let nested: Nested
}

let encoder = JSONEncoder()
let decoder = JSONDecoder()

var items: [Item] = []
items.reserveCapacity(10_000)
for i in 0..<10_000 {
    items.append(Item(
        id: i,
        name: "item_\(i)",
        value: Double(i) * 3.14159,
        tags: ["tag_\(i % 10)", "tag_\(i % 5)"],
        nested: Nested(x: i, y: i * 2)
    ))
}
let blob = try encoder.encode(items)

// Warmup
for _ in 0..<3 {
    let parsed = try decoder.decode([Item].self, from: blob)
    _ = try encoder.encode(parsed)
}

let iterations = 50
let start = DispatchTime.now()

var checksum = 0
for _ in 0..<iterations {
    let parsed = try decoder.decode([Item].self, from: blob)
    checksum += parsed.count
    let reStringified = try encoder.encode(parsed)
    checksum += reStringified.count
}

let elapsedNs = DispatchTime.now().uptimeNanoseconds - start.uptimeNanoseconds
let elapsedMs = Int(elapsedNs / 1_000_000)
print("ms:\(elapsedMs)")
print("checksum:\(checksum)")

File: docs/perry-main-src/perry-main/benchmarks/json_polyglot/bench.cpp
// JSON parse + stringify polyglot benchmark — C++ (nlohmann/json).
// 10k records, ~1 MB blob, 50 iterations.
// IDENTICAL workload to bench.ts / bench.go / bench.rs / bench.swift / bench.js.
//
// nlohmann/json is the de facto standard JSON library for C++. It's not
// the fastest (RapidJSON / simdjson are faster) but it's the most idiomatic
// and what most C++ projects reach for. Aggressive flag set in run.sh
// could swap to a faster library if we want to show the SIMD ceiling.

#include <chrono>
#include <iostream>
#include <string>
#include <vector>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

int main() {
    json items = json::array();
    for (int i = 0; i < 10000; ++i) {
        json tags = json::array();
        tags.push_back("tag_" + std::to_string(i % 10));
        tags.push_back("tag_" + std::to_string(i % 5));
        items.push_back({
            {"id", i},
            {"name", "item_" + std::to_string(i)},
            {"value", static_cast<double>(i) * 3.14159},
            {"tags", tags},
            {"nested", { {"x", i}, {"y", i * 2} }}
        });
    }
    std::string blob = items.dump();

    // Warmup
    for (int i = 0; i < 3; ++i) {
        json parsed = json::parse(blob);
        (void)parsed.dump();
    }

    constexpr int iterations = 50;
    auto start = std::chrono::steady_clock::now();

    long long checksum = 0;
    for (int iter = 0; iter < iterations; ++iter) {
        json parsed = json::parse(blob);
        checksum += static_cast<long long>(parsed.size());
        std::string reStringified = parsed.dump();
        checksum += static_cast<long long>(reStringified.size());
    }

    auto elapsed = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::steady_clock::now() - start).count();
    std::cout << "ms:" << elapsed << "\n";
    std::cout << "checksum:" << checksum << "\n";
    return 0;
}
